File: apps/mobile/lib/shared/api/tags_api.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'api_client.dart';

// ── DTO ───────────────────────────────────────────────────────────────────────

class TagDto {
  final String id;
  final String name;
  final String? color;

  const TagDto({required this.id, required this.name, this.color});

  factory TagDto.fromJson(Map<String, dynamic> j) => TagDto(
        id: j['id'] as String,
        name: j['name'] as String,
        color: j['color'] as String?,
      );
}

class TaskTagAssignment {
  final String taskId;
  final List<TagDto> tags;

  const TaskTagAssignment({required this.taskId, required this.tags});

  factory TaskTagAssignment.fromJson(Map<String, dynamic> j) => TaskTagAssignment(
        taskId: j['taskId'] as String,
        tags: (j['tags'] as List<dynamic>)
            .map((t) => TagDto.fromJson(t as Map<String, dynamic>))
            .toList(),
      );
}

// ── API ───────────────────────────────────────────────────────────────────────

class TagsApi {
  final ApiClient _client;
  const TagsApi(this._client);

  Future<List<TagDto>> list() async {
    final res = await _client.get<List<dynamic>>('/tags');
    return (res.data ?? [])
        .map((e) => TagDto.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<TaskTagAssignment>> getAllTaskAssignments() async {
    final res = await _client.get<List<dynamic>>('/tags/task-assignments');
    return (res.data ?? [])
        .map((e) => TaskTagAssignment.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<TagDto>> getTaskTags(String taskId) async {
    final res = await _client.get<List<dynamic>>('/tags/tasks/$taskId');
    return (res.data ?? [])
        .map((e) => TagDto.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<void> setTaskTags(String taskId, List<String> tagIds) async {
    await _client.put<void>('/tags/tasks/$taskId', data: {'tagIds': tagIds});
  }

  Future<TagDto> create(String name, {String? color}) async {
    final res = await _client.post<Map<String, dynamic>>('/tags', data: {
      'name': name,
      if (color != null) 'color': color,
    });
    return TagDto.fromJson(res.data!);
  }

  Future<void> delete(String id) => _client.delete<void>('/tags/$id');
}

final tagsApiProvider = Provider((ref) => TagsApi(ref.watch(apiClientProvider)));

File: apps/web/src/shared/api/tags.api.ts
import { api } from './client'

export type TagDto = {
  id: string
  userId: string
  name: string
  color: string | null
}

export const tagsApi = {
  list: () => api.get<TagDto[]>('/tags'),

  create: (name: string, color?: string | null) =>
    api.post<TagDto>('/tags', { name, color }),

  update: (id: string, patch: { name?: string; color?: string | null }) =>
    api.patch<TagDto>(`/tags/${id}`, patch),

  delete: (id: string) => api.delete<void>(`/tags/${id}`),

  getAllTaskAssignments: () =>
    api.get<{ taskId: string; tags: TagDto[] }[]>('/tags/task-assignments'),

  getTaskTags: (taskId: string) => api.get<TagDto[]>(`/tags/tasks/${taskId}`),

  setTaskTags: (taskId: string, tagIds: string[]) =>
    api.put<void>(`/tags/tasks/${taskId}`, { tagIds }),

  getAllNoteAssignments: () =>
    api.get<{ noteId: string; tags: TagDto[] }[]>('/tags/note-assignments'),

  getNoteTags: (noteId: string) => api.get<TagDto[]>(`/tags/notes/${noteId}`),

  setNoteTags: (noteId: string, tagIds: string[]) =>
    api.put<void>(`/tags/notes/${noteId}`, { tagIds }),
}
